rfcs/XXXX-gating/README.md

```diff
@@ -0,0 +1,301 @@
+```
+
+The `gating-controller` extracts the ISO8601 date from the `open.gate` annotation value,
+sets the `requestedAt` & `resetToDefaultAt`, and opens the gate for the specified window:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:00:00Z"
+  resetToDefaultAt: "2021-03-26T11:00:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:00:00Z"
+      message: "Gate scheduled for closing at 2021-03-26T11:00:00Z"
+      reason: ReconciliationSucceeded
+      status: "True"
+      type: Opened
+```
+
+While the gate is opened, all the objects that reference it are approved to reconcile at their configured interval.
+
+The SRE can decide to close the gate ahead of its schedule with:
+
+```sh
+kubectl -n flux-system annotate --overwrite gate/sre-approval \
+close.gate.fluxcd.io/requestedAt="$(date -u +"%Y-%m-%dT%H:%M:%SZ")"
+```
+
+The `gating-controller` extracts the ISO8601 date from the `close.gate` annotation value,
+compares it with the `open.gate` & `requestedAt` date and closes the gate:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:10:00Z"
+  resetToDefaultAt: "2021-03-26T10:10:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:10:00Z"
+      message: "Gate close requested"
+      reason: ReconciliationSucceeded
+      status: "False"
+      type: Opened
+```
+
+The objects that are referencing this gate, will finish their ongoing reconciliation (if any) then pause.
+
+> As a member of the SRE team, I want to block deployments in a particular time window.
+
+To enforce a maintenance window of 24 hours, you can define a `Gate` that's opened by default:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: maintenance
+  namespace: flux-system
+spec:
+  interval: 30s
+  default: opened
+  window: 24h
+```
+
+To start the maintenance window you can annotate the gate with:
+
+```sh
+kubectl -n flux-system annotate --overwrite gate/maintenance \
+close.gate.fluxcd.io/requestedAt="$(date -u +"%Y-%m-%dT%H:%M:%SZ")"
+```
+
+The `gating-controller` extracts the ISO8601 date from the `close.gate`
+annotation value and closes the gate for the specified window:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: maintenance
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:00:00Z"
+  resetToDefaultAt: "2021-03-27T10:00:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:00:00Z"
+      message: "Gate scheduled for opening at 2021-03-27T11:00:00Z"
+      reason: ReconciliationSucceeded
+      status: "False"
+      type: Opened
+```
+
+You could also schedule "No Deploy Fridays" with a CronJob that closes the `maintenance` gate at `0 0 * * FRI`.
+
```

Comment:
In addition to the time dimension, I would also propose a reference dimension, such that a gate can be defined as open to a specific reference, either in spec, or as an annotation

For example

```yaml
apiVersion: gating.toolkit.fluxcd.io/v1alpha1
kind: Gate
metadata:
  name: approved-revisions
  namespace: flux-system
spec:
  openTo: 4f08fc93e31
```

or

```yaml
apiVersion: gating.toolkit.fluxcd.io/v1alpha1
kind: Gate
metadata:
  annotations:
    open.gate.fluxcd.io/revision: 4f08fc93e31
  name: approved-revisions
  namespace: flux-system
spec:
  default: closed
```